rust/src/operations/transaction.rs

```diff
@@ -0,0 +1,175 @@
+    ) -> Self
+    where
+        T: Into<String>,
+    {
+        Self {
+            table_uri: table_uri.into(),
+            input: Arc::new(CoalescePartitionsExec::new(input)),
+            operation,
+            app_metadata,
+        }
+    }
+
+    /// Arrow schema expected to be produced by wrapped operations
+    pub fn input_operation_schema() -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+}
+
+#[async_trait]
+impl ExecutionPlan for DeltaTransactionPlan {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> ArrowSchemaRef {
+        Arc::new(OPERATION_SCHEMA.clone())
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn required_child_distribution(&self) -> Distribution {
+        Distribution::UnspecifiedDistribution
+    }
+
+    fn with_new_children(
+        &self,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    async fn execute(
+        &self,
+        _partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let mut table =
+            get_table_from_uri_without_update(self.table_uri.clone()).map_err(to_datafusion_err)?;
+
+        let data = collect(self.input.clone(), context.clone()).await?;
+        // TODO we assume that all children send a single column record batch with serialized actions
+        let actions = data
+            .iter()
+            .flat_map(|batch| deserialize_actions(batch).unwrap())
+            .collect::<Vec<_>>();
+
+        if actions.is_empty() {
+            let empty_plan = EmptyExec::new(false, self.schema());
+            return empty_plan.execute(0, context).await;
+        }
+
+        let _new_version = match table.update().await {
+            Err(_) => {
+                let mut txn = table.create_transaction(None);
+                txn.add_actions(actions);
+                let prepared_commit = txn
+                    .prepare_commit(Some(self.operation.clone()), self.app_metadata.clone())
+                    .await
+                    .map_err(to_datafusion_err)?;
+                let committed_version = table
+                    .try_commit_transaction(&prepared_commit, 0)
+                    .await
+                    .map_err(to_datafusion_err)?;
+                committed_version
+            }
+            _ => {
+                let mut txn = table.create_transaction(None);
+                txn.add_actions(actions);
+                let committed_version = txn
+                    .commit(Some(self.operation.clone()), self.app_metadata.clone())
```

Comment:
we need to commit with a specific version that the data is prepared for here, so we can trigger commit conflicts here and let the retry to resolve the conflict.